File: src/domain/entities/api_specification.py
from typing import List, Dict, Any, Optional
from src.domain.exceptions.endpoint_not_found_exception import EndpointNotFoundException

class ApiSpecification:
    def __init__(self, name: str, base_url: str, version: str, servers: List[Dict[str, Any]] = None, info: Dict[str, Any] = None):
        self.name = name
        self.base_url = base_url
        self.version = version
        self.endpoints: List[Dict[str, Any]] = []
        self.schemas: Dict[str, Any] = {}
        self.security_schemes: Dict[str, Any] = {}
        self.webhooks: Dict[str, Any] = {}
        self.servers: List[Dict[str, Any]] = servers or []
        self.tags: List[Dict[str, Any]] = []
        self.external_docs: Dict[str, Any] = {}
        self.info: Dict[str, Any] = info or {}
        self.responses: Dict[str, Any] = {}
        self.parameters: Dict[str, Any] = {}
        self.examples: Dict[str, Any] = {}
        self.request_bodies: Dict[str, Any] = {}
        self.headers: Dict[str, Any] = {}
        self.links: Dict[str, Any] = {}
        self.callbacks: Dict[str, Any] = {}
        self.path_items: Dict[str, Any] = {}
        self.security_requirements: List[Dict[str, List[str]]] = []

    def get_name(self) -> str:
        return self.name

    def get_base_url(self) -> str:
        return self.base_url

    def get_version(self) -> str:
        return self.version

    def get_endpoints(self) -> List[Dict[str, Any]]:
        return self.endpoints

    def add_endpoint(self, endpoint: Dict[str, Any]) -> None:
        self.endpoints.append(endpoint)

    def get_endpoint(self, path: str, method: str) -> Optional[Dict[str, Any]]:
        for endpoint in self.endpoints:
            if endpoint['path'] == path and endpoint['method'] == method:
                return endpoint
        return None

    def get_sample_responses_for_endpoint(self, path: str, method: str) -> Dict[str, Any]:
        endpoint = self.get_endpoint(path, method)
        if endpoint and 'responses' in endpoint:
            return endpoint['responses']
        return {}

    def get_response_by_status_code(self, path: str, method: str, status_code: str) -> Optional[Dict[str, Any]]:
        endpoint = self.get_endpoint(path, method)
        if endpoint and 'responses' in endpoint and status_code in endpoint['responses']:
            return endpoint['responses'][status_code]
        return None

    def set_schemas(self, schemas: Dict[str, Any]) -> None:
        self.schemas = schemas

    def set_security_schemes(self, security_schemes: Dict[str, Any]) -> None:
        self.security_schemes = security_schemes

    def set_webhooks(self, webhooks: Dict[str, Any]) -> None:
        self.webhooks = webhooks

    def set_servers(self, servers: List[Dict[str, Any]]) -> None:
        self.servers = servers

    def set_tags(self, tags: List[Dict[str, Any]]) -> None:
        self.tags = tags

    def set_external_docs(self, external_docs: Dict[str, Any]) -> None:
        self.external_docs = external_docs

    def set_responses(self, responses: Dict[str, Any]) -> None:
        self.responses = responses

    def set_parameters(self, parameters: Dict[str, Any]) -> None:
        self.parameters = parameters

    def set_examples(self, examples: Dict[str, Any]) -> None:
        self.examples = examples

    def set_request_bodies(self, request_bodies: Dict[str, Any]) -> None:
        self.request_bodies = request_bodies

    def set_headers(self, headers: Dict[str, Any]) -> None:
        self.headers = headers

    def set_links(self, links: Dict[str, Any]) -> None:
        self.links = links

    def set_callbacks(self, callbacks: Dict[str, Any]) -> None:
        self.callbacks = callbacks

    def set_path_items(self, path_items: Dict[str, Any]) -> None:
        self.path_items = path_items

    def set_security_requirements(self, security_requirements: List[Dict[str, List[str]]]) -> None:
        self.security_requirements = security_requirements

    def add_request_body_to_endpoint(self, path: str, method: str, request_body: Dict[str, Any]) -> None:
        self._update_endpoint(path, method, {'requestBody': request_body})

    def add_response_to_endpoint(self, path: str, method: str, status_code: str, response: Dict[str, Any]) -> None:
        self._update_endpoint(path, method, {'responses': {status_code: response}})

    def add_security_to_endpoint(self, path: str, method: str, security: List[Dict[str, List[str]]]) -> None:
        self._update_endpoint(path, method, {'security': security})

    def add_tag_to_endpoint(self, path: str, method: str, tag: str) -> None:
        endpoint = self.get_endpoint(path, method)
        if endpoint is None:
            raise EndpointNotFoundException(f"Endpoint not found for path: {path} and method: {method}")
        if 'tags' not in endpoint:
            endpoint['tags'] = []
        if tag not in endpoint['tags']:
            endpoint['tags'].append(tag)
            self._update_endpoint(path, method, {'tags': endpoint['tags']})

    def set_endpoint_summary(self, path: str, method: str, summary: str) -> None:
        self._update_endpoint(path, method, {'summary': summary})

    def set_endpoint_description(self, path: str, method: str, description: str) -> None:
        self._update_endpoint(path, method, {'description': description})

    def get_sample_responses_by_status_code(self, path: str, method: str, status_code: str) -> Dict[str, Any]:
        endpoint = self.get_endpoint(path, method)
        if endpoint is None:
            raise EndpointNotFoundException(f"Endpoint not found for path: {path} and method: {method}")

        if 'responses' not in endpoint or status_code not in endpoint['responses']:
            return {}

        response = endpoint['responses'][status_code]
        return response.get('examples', {})

    def _update_endpoint(self, path: str, method: str, data: Dict[str, Any]) -> None:
        endpoint = self.get_endpoint(path, method)
        if endpoint is None:
            raise EndpointNotFoundException(f"Endpoint not found for path: {path} and method: {method}")

        for key, value in data.items():
            if isinstance(value, dict) and key in endpoint and isinstance(endpoint[key], dict):
                endpoint[key].update(value)
            else:
                endpoint[key] = value

        self.endpoints = [e if e['path'] != path or e['method'] != method else endpoint for e in self.endpoints]

File: src/php/ApiSpecification.php
<?php

namespace TuEmpresa\LaravelAiApiIntegration\Domain\Entities;

use TuEmpresa\LaravelAiApiIntegration\Domain\Exceptions\EndpointNotFoundException;
use \cebe\openapi\spec\OpenApi;
class ApiSpecification extends OpenApi
{
    private string $name;
    private string $baseUrl;
    private array $endpoints = [];
    private array $schemas = [];
    private array $securitySchemes = [];
    private array $webhooks = [];
    private array $servers = [];
    private array $tags = [];
    private array $externalDocs = [];
    private string $version;
    private string $description = '';
    private array $info = [];
    private array $responses = [];
    private array $parameters = [];
    private array $examples = [];
    private array $requestBodies = [];
    private array $headers = [];
    private array $links = [];
    private array $callbacks = [];
    private array $pathItems = [];
    private array $securityRequirements = [];

    public function __construct(string $name, string $baseUrl, string $version, array $servers = [], array $info = []) {
        $this->name = $name;
        $this->baseUrl = $baseUrl;
        $this->version = $version;
        $this->servers = $servers;
        $this->info = $info;
    }

    public function getName(): string {
        return $this->name;
    }

    public function getBaseUrl(): string {
        return $this->baseUrl;
    }

    public function getVersion(): string {
        return $this->version;
    }

    public function getEndpoints(): array {
        return $this->endpoints;
    }

    public function addEndpoint(array $endpoint): void {
        $this->endpoints[] = $endpoint;
    }

    public function getEndpoint(string $path, string $method): ?array {
        foreach ($this->endpoints as $endpoint) {
            if ($endpoint['path'] === $path && $endpoint['method'] === $method) {
                return $endpoint;
            }
        }
        return null;
    }

    public function getSampleResponsesForEndpoint(string $path, string $method): array {
        $endpoint = $this->getEndpoint($path, $method);
        if ($endpoint && isset($endpoint['responses'])) {
            return $endpoint['responses'];
        }
        return [];
    }

    public function getResponseByStatusCode(string $path, string $method, string $statusCode): ?array {
        $endpoint = $this->getEndpoint($path, $method);
        if ($endpoint && isset($endpoint['responses'][$statusCode])) {
            return $endpoint['responses'][$statusCode];
        }
        return null;
    }

    public function getSchemas(): array {
        return $this->schemas;
    }

    public function setSchemas(array $schemas): void {
        $this->schemas = $schemas;
    }

    public function getSecuritySchemes(): array {
        return $this->securitySchemes;
    }

    public function setSecuritySchemes(array $securitySchemes): void {
        $this->securitySchemes = $securitySchemes;
    }

    public function getWebhooks(): array {
        return $this->webhooks;
    }

    public function setWebhooks(array $webhooks): void {
        $this->webhooks = $webhooks;
    }

    public function getServers(): array {
        return $this->servers;
    }

    public function setServers(array $servers): void {
        $this->servers = $servers;
    }

    public function getTags(): array {
        return $this->tags;
    }

    public function setTags(array $tags): void {
        $this->tags = $tags;
    }

    public function getExternalDocs(): array {
        return $this->externalDocs;
    }

    public function setExternalDocs(array $externalDocs): void {
        $this->externalDocs = $externalDocs;
    }

    public function getInfo(): array {
        return $this->info;
    }

    public function setResponses(array $responses): void {
        $this->responses = $responses;
    }

    public function setParameters(array $parameters): void {
        $this->parameters = $parameters;
    }

    public function setExamples(array $examples): void {
        $this->examples = $examples;
    }

    public function setRequestBodies(array $requestBodies): void {
        $this->requestBodies = $requestBodies;
    }

    public function setHeaders(array $headers): void {
        $this->headers = $headers;
    }

    public function setLinks(array $links): void {
        $this->links = $links;
    }

    public function setCallbacks(array $callbacks): void {
        $this->callbacks = $callbacks;
    }

    public function setPathItems(array $pathItems): void {
        $this->pathItems = $pathItems;
    }

    public function setSecurityRequirements(array $securityRequirements): void {
        $this->securityRequirements = $securityRequirements;
    }

    public function addRequestBodyToEndpoint(string $path, string $method, array $requestBody): void {
        $this->updateEndpoint($path, $method, ['requestBody' => $requestBody]);
    }

    public function addResponseToEndpoint(string $path, string $method, string $statusCode, array $response): void {
        $this->updateEndpoint($path, $method, ['responses' => [$statusCode => $response]]);
    }

    public function addSecurityToEndpoint(string $path, string $method, array $security): void {
        $this->updateEndpoint($path, $method, ['security' => $security]);
    }

    public function addTagToEndpoint(string $path, string $method, string $tag): void {
        $endpoint = $this->getEndpoint($path, $method);
        if ($endpoint === null) {
            throw new EndpointNotFoundException("Endpoint not found for path: $path and method: $method");
        }
        if (!in_array($tag, $endpoint['tags'])) {
            $endpoint['tags'][] = $tag;
            $this->updateEndpoint($path, $method, ['tags' => $endpoint['tags']]);
        }
    }

    public function setEndpointSummary(string $path, string $method, string $summary): void {
        $this->updateEndpoint($path, $method, ['summary' => $summary]);
    }

    public function setEndpointDescription(string $path, string $method, string $description): void {
        $this->updateEndpoint($path, $method, ['description' => $description]);
    }

    public function getSampleResponsesByStatusCode(string $path, string $method, string $statusCode): array {
        $endpoint = $this->getEndpoint($path, $method);
        if ($endpoint === null) {
            throw new EndpointNotFoundException("Endpoint not found for path: $path and method: $method");
        }

        if (!isset($endpoint['responses'][$statusCode])) {
            return [];
        }

        $response = $endpoint['responses'][$statusCode];
        return $response['examples'] ?? [];
    }

    private function updateEndpoint(string $path, string $method, array $data): void {
        $endpoint = $this->getEndpoint($path, $method);
        if ($endpoint === null) {
            throw new EndpointNotFoundException("Endpoint not found for path: $path and method: $method");
        }

        foreach ($data as $key => $value) {
            if (is_array($value) && isset($endpoint[$key]) && is_array($endpoint[$key])) {
                $endpoint[$key] = array_merge($endpoint[$key], $value);
            } else {
                $endpoint[$key] = $value;
            }
        }

        $this->endpoints = array_map(function ($e) use ($endpoint, $path, $method) {
            return ($e['path'] === $path && $e['method'] === $method) ? $endpoint : $e;
        }, $this->endpoints);
    }
}
